core/lib/object_store/src/gcs.rs

```diff
@@ -17,31 +17,39 @@ use google_cloud_storage::{
     },
 };
 use http::StatusCode;
+use rand::Rng;
 
 use crate::{
     metrics::GCS_METRICS,
     raw::{Bucket, ObjectStore, ObjectStoreError},
 };
 
-async fn retry<T, E, Fut, F>(max_retries: u16, mut f: F) -> Result<T, E>
+async fn retry<T, Fut, F>(max_retries: u16, mut f: F) -> Result<T, ObjectStoreError>
 where
-    E: fmt::Display,
-    Fut: Future<Output = Result<T, E>>,
+    Fut: Future<Output = Result<T, ObjectStoreError>>,
     F: FnMut() -> Fut,
 {
     let mut retries = 1;
-    let mut backoff = 1;
+    let mut backoff_secs = 1;
     loop {
         match f().await {
             Ok(result) => return Ok(result),
-            Err(err) => {
-                tracing::warn!(%err, "Failed GCS request {retries}/{max_retries}, retrying.");
+            Err(err) if err.is_transient() => {
```

Comment:
This change can lead to issues where `ObjectStore` previously relied on retries for logically non-transient errors (e.g., "not found" errors). Do we want to change it, or use a safer approach (e.g., consider all errors transient / retryable unless removal is concerned)?

Reply:
I think the change makes total sense. If it occurs that it breaks something (I _really_ hope it does not), we can simply revert the commit and reimplement the change with the newly obtained knowledge.